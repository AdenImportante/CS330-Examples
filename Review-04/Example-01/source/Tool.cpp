#include "Tool.h"

/**
 *
 */
Tool::Tool()
    :Item( std::string(), false )
{
    this->durability    = 0;      
    this->speed         = 0;
    this->material      = std::string();  
    this->modifier      = std::string();   
    this->modifer_level = 1;     
}

/**
 *
 */
Tool::Tool( const Tool &src )
    :Item( src.name, false )
{
    this->durability    = src.durability;      
    this->speed         = src.speed;        
    this->material      = src.material;  
    this->modifier      = src.modifier;   
    this->modifer_level = src.modifer_level;      
}

/**
 * 
 */
void Tool::display( std::ostream &outs ) const{
    outs << "  Nme: " << Item::name << "\n"
         << "  Dur: " << durability << "\n"
         << "  Spd: " << speed      << "\n"
         << "  Mtl: " << material   << "\n"
         << "  Mdr: " << modifier   << " (Lvl " << modifer_level << ")" << "\n";  
}

/**
 *
 */
void Tool::read( std::istream& ins ){
    ins >> Item::name
        >> material
        >> durability
        >> speed
        >> modifier
        >> modifer_level;
}

/**
 *
 */
Item* Tool::clone() const{
    return new Tool( *this );
}

