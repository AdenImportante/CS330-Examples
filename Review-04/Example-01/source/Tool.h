#ifndef TOOL_H_INCLUDED
#define TOOL_H_INCLUDED

#include "Item.h"

/**
 * This class represents one tool--as found in most video games. This includes
 * pickaxes and shovels.
 *
 * Tools may not be stacked. All Constructors must initialize Item::stackable
 * to false.
 */
class Tool : public Item{
    private:

    protected:
        double      durability;    ///< decreases each time the tool is used
        double      speed;         ///< base harvest/mine speed
        std::string material;      ///< material out of which the tool is composed 
        std::string modifier;      ///< one of efficiency, fortune, or unbreaking
        int         modifer_level; ///< modifier level in the range 1 to 5

    public:
        /**
         * Default to a tool with an empty name
         */
        Tool();

        /**
         * Copy Constructor
         */
        Tool( const Tool &src );

        /**
         * Retrieve tool durability
         */
        double getDurability() const;

        /**
         * Set tool durability
         */
        void setDurability( double durability );

        /**
         * Retrieve tool speed
         */
        double getSpeed() const;

        /**
         * Set tool speed
         */
        void setSpeed( double speed );

        /**
         * Retrieve tool Material
         */
        std::string getMaterial() const;

        /**
         * Set tool Material
         */
        void setMaterial( std::string m );

        /**
         * Retrieve tool Modifier
         */
        std::string getModifier() const;

        /**
         * Set tool Modifier
         */
        void setModifier( std::string m );

        /**
         * Retrieve tool Modifier Level
         */
        double getModiferLevel() const;

        /**
         * Set tool Modifier Level
         */
        void setModiferLevel( double level );

        /**
         * Print one Tool
         */
        virtual void display( std::ostream &outs ) const;

        /**
         * Read Tool attributes from an input stream
         */
        virtual void read( std::istream& ins );

        /**
         * Clone--i.e., copy--this Tool
         */
        virtual Item* clone() const;
};

/**
 *
 */
inline double Tool::getDurability() const{
    return this->durability;
}

/**
 *
 */
inline void Tool::setDurability( double durability ){
    this->durability;
}

/**
 *
 */
inline double Tool::getSpeed() const{
    return this->speed;
}

/**
 *
 */
inline void Tool::setSpeed( double speed ){
    this->speed = speed;
}

/**
 *
 */
inline std::string Tool::getMaterial() const{
    return this->material;
}

/**
 *
 */
inline void Tool::setMaterial( std::string m ){
    this->material = m;
}

/**
 *
 */
inline std::string Tool::getModifier() const{
    return this->modifier;
}

/**
 *
 */
inline void Tool::setModifier( std::string m ){
    this->modifier = m;
}

/**
 *
 */
inline double Tool::getModiferLevel() const{
    return this->modifer_level;
}

/**
 *
 */
inline void Tool::setModiferLevel( double level ){
    this->modifer_level = level;
}

#endif