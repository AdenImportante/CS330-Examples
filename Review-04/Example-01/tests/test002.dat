#! /bin/sh

# Test Tool::read and Tool::display
# formatting must match

cat <<EOF > test002.cpp

#include <iostream>
#include <iomanip>
#include <sstream>

#include "Tool.h"

using std::cout;
using std::istringstream;

int main(){
    Item *tool = new Tool(); 

    istringstream ins( "Shovel Gold 20 3 Unbreaking 2" );

    tool->read( ins );    
    tool->display( cout );  

    return 0;
}
EOF

g++ -c test002.cpp 
g++ -o test002 test002.o Tool.o Item.o

./test002