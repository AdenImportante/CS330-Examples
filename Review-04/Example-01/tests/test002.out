  Nme: Shovel
  Dur: 20
  Spd: 3
  Mtl: Gold
  Mdr: Unbreaking (Lvl 2)
